### DISCUSSION TITLE
useShallow produces error "TypeError: Iterator value X is not an entry object" under very specific conditions

### DISCUSSION THREAD

**Discussion by davefallis** [pmndrs/zustand#2794] [Bug report]
### Bug Description

I am converting a codebase to use `useShallow` for zustand v5 and ran into the error mentioned in the title. The `X` in the title depends on the value of the value of the second array item in the selector (see example).

The TypeScript playground link shows `projects` state with two items. If `projects` has exactly two items and the selector returns an array of two or more items, an exception is raised from `compareMapLike`. If `projects` is an array with <2 or >2 items _or_ you remove the `other` state from the `useShallow`, it works as expected.

Error from my projects:
```
chunk-T2SWDQEL.js?v=46bc7883:9129 Uncaught TypeError: Iterator value 1 is not an entry object
    at new Map (<anonymous>)
    at compareMapLike (zustand_shallow.js?v=46bc7883:11:55)
    at shallow (zustand_shallow.js?v=46bc7883:34:14)
    at zustand_shallow.js?v=46bc7883:66:12
    at zustand.js?v=46bc7883:38:11
    at mountSyncExternalStore (chunk-T2SWDQEL.js?v=46bc7883:11889:38)
    at Object.useSyncExternalStore (chunk-T2SWDQEL.js?v=46bc7883:12573:22)
    at Object.useSyncExternalStore (chunk-QCHXOAYK.js?v=46bc7883:1120:29)
    at useStore (zustand.js?v=46bc7883:36:38)
    at useBoundStore (zustand.js?v=46bc7883:46:39)
```

I realize the stack trace is minified. The line of `compareMapLike` that is failing is https://github.com/pmndrs/zustand/blob/0007ef43a8ef413a15f4246ffa8670a03e0499e8/src/vanilla/shallow.ts#L8

### Reproduction Link

https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgYygUwIYzXAvnAMyghDgHIAvAVwGcYMA7AEzIChRJZFa0BlACwwAbIRADu+IiXLU6jJgHoagkeLasYATzA5e9bHAC8iVnDhhiAKzTIYNAFxwGVEACM0UANoBdU3Agw-B6Ozm4ecAA+TlQirLisrMgQDHRwPHrQOMaomNgAPHpYaAB8ABSlAJRGxXClCH4WENa2DnCeAIwANHAATN6dfgFBUI7tcRUVCQRUDLbAyXAAgmBglSZmSSnwno3Ndt1DHt5GaTR8MJml6SqiYuU0VYY1njQAdLs2+3Bvh1DeEwBuBJmdAwKhQBi1PxmPLFaFmOC8KjIZBoGg0ACE8LyCjhZkm8VYaAAHpx4Ew0AQMDF4MswEA

**Comment by dbritto-dev** [comment]
@davefallis would you mind creating a minimal repro on stackblitz?

**Comment by davefallis** [comment]
https://stackblitz.com/edit/vitejs-vite-bbl6sy?file=src%2FApp.tsx

**Reply by dai-shi** [reply]
not saved?

**Reply by davefallis** [reply]
Sorry - try now

**Reply by dai-shi** [reply]
Oh, I see. I know this edge case. This is failing to detect if an iterator is from Map. Let me see how we can resolve it. Thanks for reporting.

**Reply by dai-shi** [reply]
PR 👉  #2795

**Reply by davefallis** [reply]
Something is wrong. I've created another example that shows the issue: https://stackblitz.com/edit/vitejs-vite-bwwfzi?file=src%2FApp.tsx

This example uses the zustand version from this branch - see `package.json`.

Click the button and notice the value of `activeBear` doesn't change on the screen. Swap the order of `activeBear` and `bearNames` and then click the button and you will see new value shows up.

Weirdly, I can reproduce this issue with the stock 5.0.0 release (doesn't raise an exception like in my original example), so it might be a separate issue but wanted to mention it in this context in case there is some connection.

**Reply by davefallis** [reply]
It might have something to do with the type of `activeBear`. If `activeBear` is a primitive like `number`, everything works correctly.

**Reply by davefallis** [reply]
To clarify my previous comment... If `activeBear` is a primitive like a `number`, the branch fix behaves as expected but the main 5.0.0 release blows up. It seems there is a bug in this branch **and** 5.0.0 if the `activeBear` in an object.

**Reply by dbritto-dev** [reply]
@davefallis you are returning a new array on each call, due to we use shallow comparison we are not able to compare nested objects and/or nested arrays

**Reply by dbritto-dev** [reply]
```ts
const bearNames = useBearStore(useShallow(s => s.bearNames));
const activeBear = useBearStore(s => s.activeBear);
const setActiveBear = useBearStore(s => s.setActiveBear);
```

**Reply by davefallis** [reply]
Thanks, I misunderstood the purpose of useShallow. I thought what I was doing was a valid use case for `useShallow`.

**Reply by dbritto-dev** [reply]
@davefallis don't worry, I'll add that to the troubleshooting section in the `useShallow` docs. btw, we talked about that here: https://zustand.docs.pmnd.rs/apis/shallow#troubleshooting

**Reply by LavaFields** [reply]
I might have seen https://github.com/pmndrs/zustand/discussions/1937#discussioncomment-8093370 which shows using `useShallow` to select an array of state properties from the store where one of the selected properties (`cache.posts`) is itself an array.  It has the endorsement of @dai-shi.

**Reply by dbritto-dev** [reply]
@LavaFields you need to make sure that the result of the selector is stable, with nested arrays or nested objects that doesn't happen. I know there are a few cases where you have primitive values and shallow comparison could work but I don't recommend it.

**Reply by dai-shi** [reply]
> one of the selected properties (`cache.posts`) is itself an array.

It's fine to have an array, but it will be checked with referential equality, so the array should never be mutated.
